test/blackbox-tests/test-cases/pkg/lockfile-deps-respect-constraints.t

```diff
@@ -0,0 +1,45 @@
+When populating the "deps" field of a lockfile, only packages which have locked
+versions compatible with the lockfile's package's dependency version
+constraints should be included.
+
+  $ . ./helpers.sh
+  $ mkrepo
+
+  $ mkpkg a 0.0.1
+  $ mkpkg a 0.0.2
+  $ mkpkg b
+
+The package "c" can be satisfied by b or by a.0.0.1.
+  $ mkpkg c <<EOF
+  > depends: [ "b" | "a" {= "0.0.1"} ]
+  > EOF
+
+The package "d" can only be satisfied by "a.0.0.2" which will force the package
+"c" to depend on "b" rather than "a" and ensure "a.0.0.2" is in the solution
```

Comment:
Note that `|` isn't exclusive in the constraint language. If you want to force `b` instead of `a`, you need an appropriate conflict between the two packages.

Reply:
Here I force `b` by having a second package (`d`) which depends on an incompatible version of `a`. The only way to solve for both `b` and `c` is for d to depend on `a.0.0.2` which leaves `b` as the only possible way for `a` to be satisfied.